### PR TITLE
Add custom ops for compatibility with PT Compile

### PR DESCRIPTION
**Pull request by ani300**
This PR adds basic support for torch.compile() for the non-varlen variants of Flash Attention.

This essentially allows for models that use the `flash_attn_qkvpacked_func`, `flash_attn_kvpacked_func`, and `flash_attn_func` to compile without graph breaks.

I can add unit tests if it makes sense. I'll test it with our own training pipeline for performance measurements and I'll post them later.

This uses the new custom operators API in Pytorch 2.4. I can move to the older APIs if needed, or I can look up how to make both coexist

### CODE REVIEW

**Comment by ani300** [comment]
Added varlen functions too

**Comment by tridao** [comment]
Thank you! If this requires torch 2.4, is there some way to make it optional for older pytorch version? e.g. define `torch.library.custom_op` to be a no-op? Idk what people usually do.

**Comment by ani300** [comment]
There's a different API that was used in previous pytorch versions. It's more cumbersome to use, but let me think about how to use both APIs at the same time with a version selection. I can also do a noop wrapper based on version

**Comment by ani300** [comment]
Ok, I've updated the code to only wrap things if Pytorch is 2.4 or higher. Even though we're comparing strings, `torch.__version__` implements semantic comparisons, so it should always work correctly. As soon as I have the training performance comparison I'll update the PR

**Comment by mayank31398** [comment]
aah, yikes
@ani300 I had started working on the same thing https://github.com/Dao-AILab/flash-attention/pull/1145 😓 

Ill let you handle this 😃

**Comment by ani300** [comment]
@tridao I've updated the code after testing it with SFTTrainer to make sure everything was getting called correctly and compiling without graph breaks. I've also added a check for the pytorch version so this only activates if the APIs are available.

One question I have is if it's ok to change the C++ pytorch interface so the `out`/`out_padded` padding/unpadding can be done outside of the custom op. Right now, Pytorch custom ops don't allow outputs to be aliases of other outputs, which this pair is when `head_dim % 8 == 0`. This means I have to clone one of the outputs to unalias them, which is a waste of memory bandwidth.

**Comment by tridao** [comment]
Yes let's change the C++ code to not pad the output, and have the Python part handle that.

**Comment by mayank31398** [comment]
@GLivshits I dont think it can be handled in older versions of torch

**Comment by mayank31398** [comment]
@tridao @ani300 is there any progress/updates on this?
Its a pretty neat feature to have Flash Attention fully end-to-end traceable natively.

**Comment by ani300** [comment]
Hey, I'm almost done with the C++ portion, hopefully by Monday I can update the PR and rerun all the CI to make sure it's still good with all the changes. I'll also try the suggestions from @GLivshits to Mae it work with previous versions

**Comment by ani300** [comment]
@GLivshits the way the code is currently written, the backward functions need to have `mutate_args`, or we need to do an extra memory copy on GPU, which might significantly impact performance

**Comment by ani300** [comment]
@tridao this is ready for final review and/or merging. I tested with the whole unit test suite for both pytorch 2.3 and pytorch 2.4 on an A100

**Comment by raghukiran1224** [comment]
@tridao Any update on this?

**Comment by anijain2305** [comment]
Thanks @ani300 for taking this on. I started working on this here  (https://github.com/Dao-AILab/flash-attention/pull/1209). But this is way further along. So, I closed mine.

cc @zou3519 for custom ops API usage review @Chillee

File: flash_attn/flash_attn_interface.py [hunk 1]
```diff
@@ -45,11 +46,57 @@ def _get_block_size_n(device, head_dim, is_dropout, is_causal):
...
+                return func(*args, **kwargs)
+            return inner
+        if fn is None:
+            return wrap
+        return wrap(fn)
+    _torch_custom_op_wrapper = custom_op_wrapper
+    _torch_register_fake_wrapper = register_fake_wrapper
+
+
+@_torch_custom_op_wrapper("flashattn::_flash_attn_forward", mutates_args=(), device_types="cuda")
```
Comment (anijain2305) [inline flash_attn/flash_attn_interface.py]: Nit - `s/flashattn/flash_attn/`

File: flash_attn/flash_attn_interface.py [hunk 2]
```diff
@@ -58,36 +105,71 @@ def _flash_attn_forward(
...
+    softmax_lse = torch.empty((batch_size, num_heads, seqlen_q), dtype=torch.float32, device=q.device, layout=q.layout)
+    p = torch.empty((0,), dtype=q.dtype, device=q.device, layout=q.layout)
+    if return_softmax:
+        p = torch.empty((batch_size, num_heads, round_multiple(seqlen_q, 128), round_multiple(seqlen_k, 128)), dtype=q.dtype, device=q.device, layout=q.layout)
+    rng_state = torch.empty((2,), dtype=torch.int64, device=q.device)
+
+    return out, softmax_lse, p, rng_state
+
+
+try:
```
Comment (anijain2305) [inline flash_attn/flash_attn_interface.py]: Perhaps use the same if condition

```
if torch.__version__ >= "2.4.0":
```

File: flash_attn/flash_attn_interface.py [hunk 3]
```diff
@@ -45,11 +46,57 @@ def _get_block_size_n(device, head_dim, is_dropout, is_causal):
...
+
+if torch.__version__ >= "2.4.0":
+    _torch_custom_op_wrapper = torch.library.custom_op
+    _torch_register_fake_wrapper = torch.library.register_fake
+else:
+    def custom_op_wrapper(name, fn=None, /, *, mutates_args, device_types=None, schema=None):
+        def wrap(func):
+            def inner(*args, **kwargs):
+                return func(*args, **kwargs)
+            return inner
```
Comment (anijain2305) [inline flash_attn/flash_attn_interface.py]: Is this equivalent of `return func`?

File: flash_attn/flash_attn_interface.py [hunk 4]
```diff
@@ -45,11 +46,57 @@ def _get_block_size_n(device, head_dim, is_dropout, is_causal):
...
+        if fn is None:
+            return wrap
+        return wrap(fn)
+    def register_fake_wrapper(op, fn=None, /, *, lib=None, _stacklevel=1):
+        def wrap(func):
+            def inner(*args, **kwargs):
+                return func(*args, **kwargs)
+            return inner
```
Comment (anijain2305) [inline flash_attn/flash_attn_interface.py]: Ditto - `return func`

File: flash_attn/flash_attn_interface.py [hunk 5]
```diff
@@ -45,11 +46,57 @@ def _get_block_size_n(device, head_dim, is_dropout, is_causal):
...
+        return torch_op
+    else:
+        return func
+
+
+if torch.__version__ >= "2.4.0":
+    _torch_custom_op_wrapper = torch.library.custom_op
+    _torch_register_fake_wrapper = torch.library.register_fake
+else:
+    def custom_op_wrapper(name, fn=None, /, *, mutates_args, device_types=None, schema=None):
```
Comment (anijain2305) [inline flash_attn/flash_attn_interface.py]: Nit - a more desciptive name - noop_custom_op_wrapper

And maybe a comment that we dont support < 2.4 versions.

File: flash_attn/flash_attn_interface.py [hunk 6]
```diff
@@ -45,11 +46,57 @@ def _get_block_size_n(device, head_dim, is_dropout, is_causal):
...
+        def wrap(func):
+            def inner(*args, **kwargs):
+                return func(*args, **kwargs)
+            return inner
+        if fn is None:
+            return wrap
+        return wrap(fn)
```
Comment (anijain2305) [inline flash_attn/flash_attn_interface.py]: `return fn`

**Comment by anijain2305** [comment]
Thanks @ani300 

Custom ops changes look good to me. Hope @tridao  can take a look at the other parts.

**Comment by ani300** [comment]
Tests all ran successfully again

**Comment by tridao** [comment]
Awesome, thanks!

**Comment by umarbutler** [comment]
I just built the flash-attention main branch from source and sadly I'm getting the same error message during training with the Hugging Face Trainer with `torch_compile=True`:
```
.venv/lib/python3.12/site-packages/torch/_dynamo/variables/functions.py:663: UserWarning: Graph break due to unsupported builtin flash_attn_2_cuda.PyCapsule.fwd. This function is either a Python builtin (e.g. _warnings.warn) or a third-party C/C++ Python extension (perhaps created with pybind). If it is a Python builtin, please file an issue on GitHub so the PyTorch team can add support for it and see the next case for a workaround. If it is a third-party C/C++ Python extension, please either wrap it into a PyTorch-understood custom operator (see https://pytorch.org/tutorials/advanced/custom_ops_landing_page.html for more details) or, if it is traceable, use torch.compiler.allow_in_graph.
```

**Comment by umarbutler** [comment]
It looks like I somehow ran `python setup.py bdist_wheel` and forgot I was creating a wheel (which I also ended up mistakenly deleting moments ago) 😅 hopefully second time's the charm

**Comment by ani300** [comment]
I was worried for a second, as I just used it with the Huggingface Trainer yesterday successfully 😥

**Comment by umarbutler** [comment]
@ani300 I can confirm that it works! And much faster than PyTorch SDPA, shaved off 50 training hours (400h -> 350h) 😆

**Comment by vadimkantorov** [comment]
Does it also work for fullgraph?

Thanks!
